### PR TITLE
Make SyncEngine an interface.

### PR DESCRIPTION
**Pull request by wu-hui**

### CODE REVIEW

File: packages/firestore/src/core/sync_engine.ts [hunk 1]
```diff
@@ -912,13 +957,46 @@ export class SyncEngine implements RemoteSyncer {
...
+    datastore,
+    sharedClientState,
+    currentUser,
+    maxConcurrentLimboResolutions
+  );
+}
+
 /**
- * An impplementation of SyncEngine that implement SharedClientStateSyncer for
+ * An extension of SyncEngine that also include SharedClientStateSyncer for
```
Comment (schmidt-sebastian) [inline packages/firestore/src/core/sync_engine.ts:980]: s/include/includes

File: packages/firestore/src/core/sync_engine.ts [hunk 2]
```diff
@@ -145,7 +145,94 @@ export interface SyncEngineListener {
...
+    source: OnlineStateSource
+  ): void;
+
+  /**
+   * Registers a user callback that resolves when all pending mutations at the moment of calling
+   * are acknowledged .
+   */
+  registerPendingWritesCallback(callback: Deferred<void>): Promise<void>;
+
+  activeLimboDocumentResolutions(): SortedMap<DocumentKey, TargetId>;
```
Comment (schmidt-sebastian) [inline packages/firestore/src/core/sync_engine.ts:214]: Add `// Visible for testing`

File: packages/firestore/src/core/sync_engine.ts [hunk 3]
```diff
@@ -145,7 +145,94 @@ export interface SyncEngineListener {
...
+
+  enqueuedLimboDocumentResolutions(): DocumentKey[];
```
Comment (schmidt-sebastian) [inline packages/firestore/src/core/sync_engine.ts:216]: Add // Visible for testing

File: packages/firestore/src/core/sync_engine.ts (hunk 2 above)
Comment (wu-hui) [reply packages/firestore/src/core/sync_engine.ts:215]: Done.

File: packages/firestore/src/core/sync_engine.ts (hunk 3 above)
Comment (wu-hui) [reply packages/firestore/src/core/sync_engine.ts:218]: Done.

File: packages/firestore/src/core/sync_engine.ts (hunk 1 above)
Comment (wu-hui) [reply packages/firestore/src/core/sync_engine.ts]: Done.

**Pull request merged**